File: app/src/main/java/com/jodelXposed/hooks/UniqueDeviceIdentifierStuff.java
package com.jodelXposed.hooks;

import com.jodelXposed.models.Hookvalues;
import com.jodelXposed.models.UDI;
import com.jodelXposed.utils.Options;

import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.callbacks.XC_LoadPackage;

import static com.jodelXposed.utils.Log.xlog;
import static de.robv.android.xposed.XposedHelpers.findAndHookMethod;

public class UniqueDeviceIdentifierStuff {

    /**
     * Spoof UID
     */
    public UniqueDeviceIdentifierStuff(XC_LoadPackage.LoadPackageParam lpparam) {
        Hookvalues hooks = Options.INSTANCE.getHooks();
        findAndHookMethod(hooks.Class_UniqueDeviceIdentifier, lpparam.classLoader, hooks.UDI_GetUdiMethod, new XC_MethodHook() {
            @Override
            protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                String realUDI = (String) param.getResult();
                UDI udiOptions = Options.INSTANCE.getUdi();
                if (udiOptions.getUdi().length() == 0) {
                    udiOptions.setUdi(realUDI);
                    xlog("Backing up UDI");
                    Options.INSTANCE.save();
                }
                if (udiOptions.getActive()) {
                    if (udiOptions.getUdi().equals(realUDI) || udiOptions.getUdi().length() == 0) {
                        xlog("Using real UDI");
                    } else {
                        String spoofUDI = udiOptions.getUdi();
                        xlog("UDI spoof = " + spoofUDI);
                        param.setResult(spoofUDI);
                    }
                }
            }
        });
    }
}

File: app/src/main/java/com/jodelXposed/hooks/UniqueDeviceIdentifierStuff.kt
package com.jodelXposed.hooks

import com.jodelXposed.utils.Options

import de.robv.android.xposed.XC_MethodHook
import de.robv.android.xposed.callbacks.XC_LoadPackage.LoadPackageParam

import com.jodelXposed.utils.Log.dlog
import de.robv.android.xposed.XposedHelpers.findAndHookMethod

class UniqueDeviceIdentifierStuff(lpparam: LoadPackageParam) {
    init {
        findAndHookMethod(Options.hooks.Class_UniqueDeviceIdentifier, lpparam.classLoader, Options.hooks.Method_UDI_GetUdiMethod, object : XC_MethodHook() {
            @Throws(Throwable::class)
            override fun afterHookedMethod(param: XC_MethodHook.MethodHookParam) {
                val realUDI = param.result as String
                if (Options.udi.udi.length == 0) {
                    Options.udi.udi = realUDI
                    dlog("Backing up UDI")
                    Options.save()
                    return
                }

                dlog("UDI spoof = ${Options.udi.udi}")
                param.result = Options.udi.udi
            }
        })
    }
}
